src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseParticipationService.java

```diff
@@ -145,39 +146,39 @@ public ProgrammingExerciseParticipation findProgrammingExerciseParticipationWith
         }
         return false;
     }
 
-    private boolean canAccessParticipation(@NotNull ProgrammingExerciseStudentParticipation participation) {
-        var user = userRepository.getUserWithGroupsAndAuthorities();
-        return participation.isOwnedBy(user) || authCheckService.isAtLeastTeachingAssistantForExercise(participation.getExercise(), user);
-    }
-
-    private boolean canAccessParticipation(@NotNull SolutionProgrammingExerciseParticipation participation) {
-        // Note: if this participation was retrieved as Participation (abstract super class) from the database, the programming exercise might not be correctly initialized
-        // To prevent null pointer exceptions, we therefore retrieve it again as concrete solution programming exercise participation
-        if (participation.getProgrammingExercise() == null || !Hibernate.isInitialized(participation.getProgrammingExercise())) {
-            participation.setProgrammingExercise(solutionParticipationRepository.findById(participation.getId()).get().getProgrammingExercise());
-        }
-        return authCheckService.isAtLeastTeachingAssistantForExercise(participation.getExercise(), null);
-    }
-
-    private boolean canAccessParticipation(@NotNull TemplateProgrammingExerciseParticipation participation) {
+    /**
+     * Returns whether a user is allowed to access a given participation.
+     *
+     * @param <T>           The {@link ProgrammingExerciseParticipation} sub-class
+     * @param participation A participation of type <code>T</code>, must not be null
+     * @param repository    The database repository where participations of type <code>T</code> reside in
+     * @param user          The user, may be null, in which case the current user is fetched and used.
+     * @return <code>true</code> if the current user is allowed to access the given participation, <code>false</code> otherwise
+     */
+    private <T extends ProgrammingExerciseParticipation> boolean canAccessParticipation(@NotNull T participation, JpaRepository<T, Long> repository, User user) {
         // Note: if this participation was retrieved as Participation (abstract super class) from the database, the programming exercise might not be correctly initialized
-        // To prevent null pointer exceptions, we therefore retrieve it again as concrete template programming exercise participation
+        // To prevent null pointer exceptions, we therefore retrieve it again as concrete sub-class instance by using the provided repository
         if (participation.getProgrammingExercise() == null || !Hibernate.isInitialized(participation.getProgrammingExercise())) {
-            participation.setProgrammingExercise(templateParticipationRepository.findById(participation.getId()).get().getProgrammingExercise());
+            participation.setProgrammingExercise(repository.findById(participation.getId()).get().getProgrammingExercise());
```

Comment:
I think we should avoid refetching the data from the database if this was not an issue before